codalab/worker/torque_bundle_manager.py

```diff
@@ -0,0 +1,213 @@
+import logging
+import os
+import re
+import subprocess
+import time
+import traceback
+
+from codalab.common import State
+from codalab.lib import path_util
+from codalab.worker.bundle_manager import BundleManager
+from codalab.worker.worker_info_accessor import WorkerInfoAccessor
+from worker.file_util import remove_path
+
+
+logger = logging.getLogger(__name__)
+
+
+class TorqueBundleManager(BundleManager):
+    def __init__(self, codalab_manager, torque_config):
+        assert(codalab_manager.worker_model().shared_file_system)
+        self._torque_ssh_host = torque_config['ssh_host']
+        self._torque_bundle_service_url = torque_config['bundle_service_url']
+        self._torque_password_file = torque_config['password_file']
+        self._torque_log_dir = torque_config['log_dir']
+        path_util.make_directory(self._torque_log_dir)
+        if 'worker_code_dir' in torque_config:
+            self._torque_worker_code_dir = torque_config['worker_code_dir']
+        else:
+            codalab_cli = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+            self._torque_worker_code_dir = os.path.join(codalab_cli, 'worker')
+
+    def _schedule_run_bundles(self):
+        """
+        This method implements a state machine. The states are:
+
+        Staged, no job_handle, no worker, no worker_run DB entry:
+            Ready to run on a CodaLab-owned Torque worker or on a user-owned
+            worker.
+
+        For user-owned workers, the states are the same as used in the
+        DefaultBundleManager:
+
+        Starting, no job_handle, has worker_run DB entry:
```

Comment:
to make it clearer, change to 'STARTING: no job_handle ...' (same for all other states).